lib/live_admin/resource.ex

```diff
@@ -198,29 +198,41 @@ defmodule LiveAdmin.Resource do
 
   def fields(resource, config) do
     schema = Keyword.fetch!(resource.__live_admin_config__(), :schema)
+    hidden_fields = LiveAdmin.fetch_config(resource, :hidden_fields, config)
+    immutable_fields = LiveAdmin.fetch_config(resource, :immutable_fields, config)
 
-    Enum.flat_map(schema.__schema__(:fields), fn field_name ->
-      resource
-      |> LiveAdmin.fetch_config(:hidden_fields, config)
-      |> Enum.member?(field_name)
-      |> case do
-        false ->
-          [
-            {field_name, schema.__schema__(:type, field_name),
-             [
-               immutable:
-                 resource
-                 |> LiveAdmin.fetch_config(:immutable_fields, config)
-                 |> Enum.member?(field_name)
-             ]}
-          ]
-
-        true ->
-          []
-      end
+    schema.__schema__(:fields)
+    |> Enum.reject(&(&1 in hidden_fields))
+    |> Enum.map(fn field_name ->
+      type = schema.__schema__(:type, field_name)
+      {known_type, is_custom_type?} = parse_type(type)
```

Comment:
can you name this function something a bit more specific, like `extract_native_type`

`known` is a bit vague and I'm not sure it's accurate since LA doesn't know what will be here at this point. docs seem to refer to that data as "native"

pretty sure it could just return the exported native type `nil`

Reply:
I do want to defend the function name because I saw it as filling this role

https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/

We are parsing from something unknown to something known.

I don't follow this bit

> pretty sure it could just return the exported native type or nil and config can just be type || native_type

Reply:
No validating is being done here, it needs to handle every possible type Ecto can throw at us, they are just handled differently. I'm ok with this, but I'm a bit bothered by assigning to `native_type` since it can actually return other types as well, viz. embeds. But I think that is mostly because this is really mixing concerns a bit and this logic really just belongs in the component rendering logic. But I can always refactor that later.